rules/epmark.go

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) 2018 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rules
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/projectcalico/felix/markbits"
+	log "github.com/sirupsen/logrus"
+)
+
+// Endpoint Mark Mapper (EPM) provides set of functions to manage allocation/free endpoint mark bit
+// given a mark bit mask.
+type EndpointMarkMapper interface {
+	GetEndpointMark(ep string) (uint32, error)
+	RemoveEndpointMark(ep string)
+	SetEndpointMark(ep string, mark uint32) (error)
+}
+
+type DefaultEPMarkManager struct {
+	markBitsManager *markbits.MarkBitsManager
+	maxPosition     int
+
+	activeEndpointToPosition map[string]int
+	activePositionToEndpoint map[int]string
+
+	mutex sync.Mutex
+}
+
+func NewEndpointMarkMapper(markMask uint32) EndpointMarkMapper {
+	markBitsManager := markbits.NewMarkBitsManager(markMask, "endpoint-iptable-mark")
+
+	return &DefaultEPMarkManager{
+		markBitsManager:          markBitsManager,
+		maxPosition:              markBitsManager.CurrentFreeNumberOfMark(), // This includes zero
+		activeEndpointToPosition: map[string]int{},
+		activePositionToEndpoint: map[int]string{},
+	}
+}
+
+func (epmm *DefaultEPMarkManager) GetEndpointMark(ep string) (uint32, error) {
+	length := len(ep)
+	if length == 0 {
+		return 0, errors.New("Invalid endpoint name")
+	}
+
+	epmm.mutex.Lock()
+	defer epmm.mutex.Unlock()
+
+	// Return current mark for Endpoint if it already has one.
+	if pos, ok := epmm.activeEndpointToPosition[ep]; ok {
+		mark, err := epmm.markBitsManager.MapNumberToMark(pos)
+		if err != nil {
+			return 0, err
+		}
+		return mark, nil
+	}
+
+	// Try to allocate a position based on a simple hash from endpoint name.
+	// Make sure we are likely to get a good performance based on average 10 to 50 pods per node.
+
+	// Take last two bytes or one byte from name and take the modulus of max position.
+	var total int
```

Comment:
I suggest swapping this for an off-the-shelf hash function.  Go has a package built in with FNV32, which is a good, fast hash function.  Something like ```
h := fnv.New32()
h.Write([]byte(ep))
prospect := 1 + (h.Sum32() % (epmm.maxPosition - 1))
```

Reply:
Agree.